scrapy/utils/log.py

```diff
@@ -0,0 +1,155 @@
+
+import scrapy
+from scrapy.settings import overridden_settings
+
+logger = logging.getLogger(__name__)
+
+
+class FailureFormatter(logging.Filter):
+    """Extract exc_info from Failure instances provided as contextual data
+
+    This filter mimics Twisted log.err formatting for its first `_stuff`
+    argument, which means that reprs of non Failure objects are appended to the
+    log messages.
+    """
+
+    def filter(self, record):
+        failure = record.__dict__.get('failure')
+        if failure:
+            if isinstance(failure, Failure):
+                record.exc_info = (failure.type, failure.value, failure.tb)
+            else:
+                record.msg += os.linesep + repr(failure)
+        return True
+
+
+class TopLevelFormatter(logging.Filter):
+    """Keep only top level loggers's name (direct children from root) from
+    records.
+
+    This filter will replace Scrapy loggers' names with 'scrapy'. This mimics
+    the old Scrapy log behaviour and helps shortening long names.
+
+    Since it can't be set for just one logger (it won't propagate for its
+    children), it's going to be set in the root handler, with a parametrized
+    `loggers` list where it should act.
+    """
+
+    def __init__(self, loggers=None):
+        self.loggers = loggers or []
+
+    def filter(self, record):
+        if any(record.name.startswith(l + '.') for l in self.loggers):
+            record.name = record.name.split('.', 1)[0]
+        return True
+
+
+DEFAULT_LOGGING = {
+    'version': 1,
+    'disable_existing_loggers': False,
+    'filters': {
+        'failure_formatter': {
+            '()': 'scrapy.utils.log.FailureFormatter',
+        },
+    },
+    'loggers': {
+        'scrapy': {
+            'level': 'DEBUG',
+            'filters': ['failure_formatter'],
+        },
+        'twisted': {
+            'level': 'ERROR',
+        },
+    }
+}
+
+
+def configure_logging(settings=None):
+    """Initialize and configure default loggers
+
+    This function does:
+      - Route warnings and twisted logging through Python standard logging
+      - Set FailureFormatter filter on Scrapy logger
+      - Assign DEBUG and ERROR level to Scrapy and Twisted loggers respectively
+      - Create a handler for the root logger according to given settings
+    """
+    if not sys.warnoptions:
+        # Route warnings through python logging
+        logging.captureWarnings(True)
+
+    observer = twisted_log.PythonLoggingObserver('twisted')
+    observer.start()
+
+    dictConfig(DEFAULT_LOGGING)
+    if settings:
+        logging.root.setLevel(logging.NOTSET)
+
+        if settings.getbool('LOG_STDOUT'):
+            sys.stdout = StreamLogger(logging.getLogger('stdout'))
+
+        # Set up the default log handler
+        filename = settings.get('LOG_FILE')
+        if filename:
+            encoding = settings.get('LOG_ENCODING')
+            handler = logging.FileHandler(filename, encoding=encoding)
```

Comment:
@Curita I personally vote for closing file handler explicitly - it's best practice. Leaving file handler opened and relying on the fact that it will be closed on interpreter exit make code buggy and hard to reuse in other projects. For example https://github.com/scrapinghub/scrapyrt/pull/14 wouldn't have existed if someone had thought about this while implementing current scrapy logging